### PR TITLE
CCM upgrade flow skeleton for freeipa service

### PR DESCRIPTION
**Pull request by afarsang**
See detailed description in the commit message.

### CODE REVIEW

**Review by bergerdenes — CHANGES_REQUESTED** [review]
minor nit

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+import static com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.DetailedStackStatus.UPGRADE_CCM_IN_PROGRESS;
+
+import javax.inject.Inject;
+
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.freeipa.api.v1.freeipa.stack.model.common.DetailedStackStatus;
+import com.sequenceiq.freeipa.service.stack.StackUpdater;
+
+@Component
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:14]: nit: `@Service`

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
+package com.sequenceiq.freeipa.flow.stack.upgrade.ccm;
+
+import com.sequenceiq.flow.core.FlowState;
+import com.sequenceiq.flow.core.RestartAction;
+import com.sequenceiq.freeipa.flow.FillInMemoryStateStoreRestartAction;
+
+public class UpgradeCcmState implements FlowState {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java:7]: could you use an enum as we usually do?

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+    public Action<?, ?> pushSaltStates() {
+        return new AbstractUpgradeCcmEventAction() {
+            @Override
+            protected void doExecute(UpgradeCcmContext context, UpgradeCcmEvent payload, Map<Object, Object> variables) {
+                LOGGER.info("Starting pushing salt states for CCM upgrade {}", payload);
+                upgradeCcmService.pushSaltStatesState(context.getStack().getId());
+                sendEvent(context, UPGRADE_CCM_PUSH_SALT_STATES_EVENT.createBasedOn(payload));
+            }
+        };
+    }
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:90]: I'm a bit confused what is the flow of this process is, maybe it's related to the fact that there is no description to the commit which has the expected logic details.
So what is the plane to make the changes on the host? Are you planning a full salt state upgrade which would first push down CCMV2 related configs and start it, then would come the CP rereg+healtcheck, then once again a salt state upgrade to disable old CCM.
Or are you planning to play with pushing down single states/pillars?

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    public void setCcmConnectivityParameters(CcmConnectivityParameters ccmConnectivityParameters) {
+        this.ccmConnectivityParameters = ccmConnectivityParameters;
+    }
+
+    @Override
+    public String toString() {
+        return "UpgradeCcmEvent{" +
+                "selector=" + selector +
+                ", oldTunnel=" + oldTunnel +
+                ", ccmConnectivityParameter='" + ccmConnectivityParameters + '\'' +
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:54]: this class doesn't have `toString` implementation

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/AbstractUpgradeCcmEventHandler.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package com.sequenceiq.freeipa.flow.stack.upgrade.ccm.handler;
+
+import com.sequenceiq.flow.reactor.api.handler.ExceptionCatcherEventHandler;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmEvent;
+
+abstract class AbstractUpgradeCcmEventHandler extends ExceptionCatcherEventHandler<UpgradeCcmEvent> {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/AbstractUpgradeCcmEventHandler.java:6]: why is this necessary?

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmChangeTunnelHandler.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+import com.sequenceiq.cloudbreak.common.event.Selectable;
+import com.sequenceiq.flow.reactor.api.handler.HandlerEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.UpgradeCcmService;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmEvent;
+import com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event.UpgradeCcmFailureEvent;
+
+import reactor.bus.Event;
+
+@Component
+public class UpgradeCcmChangeTunnelHandler extends AbstractUpgradeCcmEventHandler {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmChangeTunnelHandler.java:22]: are you planning to add unit tests to the handler later?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaUpgradeCcmService.java [hunk 7]
```diff
@@ -64,11 +68,14 @@ private void validateCcmUpgrade(Stack stack) {
...
         try {
-            LOGGER.info("Starting CCM upgrade flow");
-            // TODO log & update stack status (if needed), then send event to FreeIpaFlowManager; see CB-14571
-            // TODO what to do with the FlowIdentifier?
+            LOGGER.info("Starting CCM upgrade flow chain, new status: {}", UPGRADE_CCM_REQUESTED);
+            stackUpdater.updateStackStatus(stack, UPGRADE_CCM_REQUESTED, "Starting of CCM upgrade requested");
+            UpgradeCcmFlowChainTriggerEvent event = new UpgradeCcmFlowChainTriggerEvent(UPGRADE_CCM_CHAIN_TRIGGER_EVENT,
+                    operation.getOperationId(), stack.getId());
+            flowManager.notify(UPGRADE_CCM_CHAIN_TRIGGER_EVENT, event);
+            // Ignore returned FlowIdentifier
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaUpgradeCcmService.java:78]: ?

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java (hunk 2 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java:7]: I can but I wouldn't. With this I was able to use literals to name the action beans. Enum had no addition value in the flow. but you can convince me.

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmChangeTunnelHandler.java (hunk 6 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/UpgradeCcmChangeTunnelHandler.java:22]: yes.
Currently it has an integration test implemented as a unit test

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaUpgradeCcmService.java (hunk 7 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaUpgradeCcmService.java:78]: I'll remove the comment.

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java (hunk 1 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:14]: ok, I am going to fix it.

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/AbstractUpgradeCcmEventHandler.java (hunk 5 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/handler/AbstractUpgradeCcmEventHandler.java:6]: Fixing a type parameter

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java (hunk 2 above)
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java:7]: I see, but if this would be an enum, you could do it with the other way around, like:
```
public static final String UPGRADE_CCM_CHECK_PREREQUISITES_STATE_NAME = UPGRADE_CCM_CHECK_PREREQUISITES_STATE.name();
```
what do you think?

File: freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/DetailedStackStatus.java [hunk 8]
```diff
@@ -60,19 +60,25 @@
...
     DELETED_ON_PROVIDER_SIDE(Status.DELETED_ON_PROVIDER_SIDE, AvailabilityStatus.UNAVAILABLE),
     UNHEALTHY(Status.UNHEALTHY, AvailabilityStatus.AVAILABLE),
+    // Salt state update statuses
     SALT_STATE_UPDATE_IN_PROGRESS(Status.UPDATE_IN_PROGRESS, AvailabilityStatus.AVAILABLE),
-    SALT_STATE_UPDATE_FAILED(Status.UPDATE_FAILED, AvailabilityStatus.AVAILABLE);
+    SALT_STATE_UPDATE_FAILED(Status.UPDATE_FAILED, AvailabilityStatus.AVAILABLE),
+    // CCM upgrade statuses
+    UPGRADE_CCM_REQUESTED(Status.UPGRADE_CCM_REQUESTED, AvailabilityStatus.AVAILABLE),
+    UPGRADE_CCM_IN_PROGRESS(Status.UPGRADE_CCM_IN_PROGRESS, AvailabilityStatus.AVAILABLE),
+    UPGRADE_CCM_FAILED(Status.UPGRADE_CCM_FAILED, AvailabilityStatus.UNAVAILABLE);
```
[the remark this answers is not in the document]
Comment (bergerdenes) [reply freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/DetailedStackStatus.java:81]: agree

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java (hunk 3 above)
Comment (bergerdenes) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:90]: pillar update (with top.sls) and then bootstrap salt (to reload pillar with updated top.sls and propagate new states)

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+        testFlow(CALLED_ONCE_TILL_GENERATE_USERDATE, true, false);
+    }
+
+    @Test
+    public void testCcmUpgradeWhenRemoveMinaFailsInChain() throws Exception {
+        doThrow(new BadRequestException()).when(upgradeCcmService).removeMina(STACK_ID);
+        testFlow(CALLED_ONCE_TILL_REMOVE_MINA, false, false);
+    }
+
+    private void testFlow(int calledOnceTillUpgrade, boolean ccmUpgradeSuccess, boolean userDataUpdateSuccess) throws Exception {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:166]: `calledOnceTillUpgrade` -> `calledOnceCount`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+        verify(resourcesApi, times(userDataUpdateSuccess ? 1 : 0)).updateUserData(any(), any(), any(), eq(USER_DATA));
+
+        verify(upgradeCcmService, times(ccmUpgradeSuccess ? 0 : 1)).failedState(STACK_ID);
+
+        verify(operationService, times(ccmUpgradeSuccess && userDataUpdateSuccess ? 1 : 0)).completeOperation(any(), any(), any(), any());
+        verify(operationService, times(ccmUpgradeSuccess && userDataUpdateSuccess ? 0 : 1)).failOperation(any(), any(), any());
+
+    }
+
+    private void verifyServiceCalls(int tillInt) throws Exception {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:186]: `tillInt` -> `calledOnceCount`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+        verify(upgradeCcmService, times(expected[i++])).healthCheck(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).removeMinaState(STACK_ID);
+        verify(upgradeCcmService, times(expected[i++])).removeMina(STACK_ID);
+        verify(userDataService, times(expected[i++])).createUserData(STACK_ID);
+        verify(resourcesApi, times(expected[i++])).updateUserData(any(), any(), any(), eq(USER_DATA));
+
+
+    }
+
+    private void flowFinishedSuccessfully(int numberOfFlows) {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:212]: `numberOfFlows` -> `numberOfSuccessfulFlows`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+    private static final String USER_DATA = "hello hello is there anybody out there";
+
+    private static final long STACK_ID = 1L;
+
+    private static final int ALL_CALLED_ONCE = 20;
+
+    private static final int CALLED_ONCE_TILL_REMOVE_MINA = 16;
+
+    private static final int CALLED_ONCE_TILL_GENERATE_USERDATE = 17;
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:86]: `USERDATE` -> `USERDATA`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,200 @@
...
+@ExtendWith(SpringExtension.class)
+class UpdateUserDataFlowIntegrationTest {
+
+    private static final String USER_CRN = "crn:cdp:iam:us-west-1:" + UUID.randomUUID() + ":user:" + UUID.randomUUID();
+
+    private static final long STACK_ID = 1L;
+
+    private static final int ALL_CALLED_ONCE = 1;
+
+    private static final int CALLED_ONCE_TILL_GENERATE_USERDATE = 1;
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java:69]: `USERDATE` -> `USERDATA`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java [hunk 14]
```diff
@@ -0,0 +1,200 @@
...
+        testFlow(ALL_CALLED_ONCE, true);
+    }
+
+    @Test
+    public void testUserDataUpdateWhenNewUserDataFails() throws Exception {
+        doThrow(new BadRequestException()).when(userDataService).createUserData(STACK_ID);
+        testFlow(CALLED_ONCE_TILL_GENERATE_USERDATE, false);
+    }
+
+    private void testFlow(int calledOnceTillUpgrade, boolean b) throws Exception {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java:142]: `calledOnceTillUpgrade` -> `calledOnceCount`.

Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java:142]: `b` -> `success`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java [hunk 15]
```diff
@@ -0,0 +1,200 @@
...
+    }
+
+    private void verifyFinishingStatCalls(boolean success) throws Exception {
+        verify(resourcesApi, times(success ? 1 : 0)).updateUserData(any(), any(), any(), eq(USER_DATA));
+        verify(operationService, times(success ? 1 : 0)).completeOperation(any(), any(), any(), any());
+        verify(operationService, times(success ? 0 : 1)).failOperation(any(), any(), any());
+
+    }
+
+    private void verifyServiceCalls(int tillInt) {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/update/UpdateUserDataFlowIntegrationTest.java:158]: `tillInt` -> `calledOnceCount`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+        testFlow(CALLED_ONCE_TILL_HEALTH_CHECK, false);
+    }
+
+    @Test
+    public void testCcmUpgradeWhenRemoveMinaFails() {
+        doThrow(new BadRequestException()).when(upgradeCcmService).removeMina(1L);
+        testFlow(CALLED_ONCE_TILL_REMOVE_MINA, false);
+    }
+
+    private void testFlow(int calledOnceTillUpgrade, boolean b) {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:154]: `b` -> `success`.

Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:154]: `calledOnceTillUpgrade` -> `calledOnceCount`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+
+    private void verifyFinishingStatCalls(boolean success) {
+        verify(upgradeCcmService, times(success ? 1 : 0)).finishedState(STACK_ID);
+        verify(operationService, times(success ? 1 : 0)).completeOperation(any(), any(), any(), any());
+        verify(upgradeCcmService, times(success ? 0 : 1)).failedState(STACK_ID);
+        verify(operationService, times(success ? 0 : 1)).failOperation(any(), any(), any());
+
+    }
+
+    private void verifyServiceCalls(int tillInt) {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:171]: `tillInt` -> `calledOnceCount`.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+
+@ActiveProfiles("integration-test")
+@ExtendWith(SpringExtension.class)
+class UpgradeCcmFlowIntegrationTest {
+
+    private static final String USER_CRN = "crn:cdp:iam:us-west-1:" + UUID.randomUUID() + ":user:" + UUID.randomUUID();
+
+    private static final long STACK_ID = 1L;
+
+    private static final int ALL_CALLED_ONCE = 20;
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmFlowIntegrationTest.java:58]: Not a big deal but I think this should be 16.

**Review by lajosrodek — CHANGES_REQUESTED** [review]

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java (hunk 11 above)
Comment (afarsang) [reply freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:212]: not necessary successful, rather running. (if first failed, then it has 2 records, but the second flow won't start)

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java (hunk 4 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:54]: I don't get this unfortunately. can you elaborate what you mean?

Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:54]: `CcmConnectivityParameters` doesn't implement `toString` method. Logging it here doesn't make any sense in that case (no useful information would be logged). So either you should implement `toString` in that class too, or you can remove logging it from here. Which makes more sense is up to you.

Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:54]: right, thx

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java (hunk 2 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java:7]: unfortunately this is not work:
`public static final String UPGRADE_CCM_CHECK_PREREQUISITES_STATE_NAME = UPGRADE_CCM_CHECK_PREREQUISITES_STATE.name();`
since this is not resulting as constant, and I can't use it in annotation

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java (hunk 3 above)
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:90]: in that case it seems this should be a flowchain and the salt state update flow should be called

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java (hunk 2 above)
Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmState.java:7]: alright, tanks for checking. if you find this better I can accept. although it's a bit different of how we usually implement flows which might cause a bit of confusion later

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+        stackUpdater.updateStackStatus(stackId, detailedStatus, statusReason);
+    }
+
+    public void failedState(Long stackId) {
+        DetailedStackStatus detailedStatus = UPGRADE_CCM_FAILED;
+        String statusReason = "Upgrade CCM failed";
+        stackUpdater.updateStackStatus(stackId, detailedStatus, statusReason);
+    }
+
+    public void changeTunnelState(long stackId) {
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:42]: these all should be `Long`

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
+package com.sequenceiq.freeipa.flow.stack.upgrade.ccm.event;
+
+import java.util.Optional;
+
+import com.sequenceiq.cloudbreak.ccm.cloudinit.CcmConnectivityParameters;
+import com.sequenceiq.cloudbreak.clusterproxy.Tunnel;
+import com.sequenceiq.freeipa.flow.stack.StackEvent;
+
+public class UpgradeCcmEvent extends StackEvent {
+    private String selector;
```
Comment (lacikaaa) [inline freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:10]: `StackEvent` already has `selector`

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java (hunk 19 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/UpgradeCcmService.java:42]: ok

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java (hunk 20 above)
Comment (afarsang) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/event/UpgradeCcmEvent.java:10]: right

File: freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java (hunk 3 above)
Comment (bergerdenes) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:90]: Unfortunately the salt update flow is not prepared to receive payload from external object in OrchestratorConfigHandler
on the other hand, this flow also installs FreeIPA which is not desired in this case.
It would mean a bit more hack than the currently created customized flow in this PR.

Comment (lacikaaa) [reply freeipa/src/main/java/com/sequenceiq/freeipa/flow/stack/upgrade/ccm/action/UpgradeCcmActions.java:90]: I'm still confused, the salt stuff should be idempotent, you should be able to run highstate at any time on a machine or the whole cluster and it should be fine.
Anything which is salt level logic should be controlled by pillars and grains. So you should be able to push down all the necessary states and pillars for the ccm upgrade using the flow. Yes, this will invoke a highstate, but that shouldn't be a problem at all, you might set the necessary grains before and everything should be in a good state after the highstate. Or you can add the grains later and call a state or highstate after that.

>  installs FreeIPA

I don't get it, FreeIPA is already installed, it won't be reinstalled (at least it shouldn't be reinstalled, if it is, then it's a bug)

**Review by lacikaaa — APPROVED** [review]

**Review by bergerdenes — APPROVED** [review]

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java (hunk 11 above)
Comment (lajosrodek) [reply freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java]: Oh, right. Makes sense.

File: freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java [hunk 21]
```diff
@@ -0,0 +1,259 @@
...
+        testFlow(CALLED_ONCE_TILL_GENERATE_USERDATA, true, false);
+    }
+
+    @Test
+    public void testCcmUpgradeWhenRemoveMinaFailsInChain() throws Exception {
+        doThrow(new BadRequestException()).when(upgradeCcmService).removeMina(STACK_ID);
+        testFlow(CALLED_ONCE_TILL_REMOVE_MINA, false, false);
+    }
+
+    private void testFlow(int calledOnce, boolean ccmUpgradeSuccess, boolean userDataUpdateSuccess) throws Exception {
```
Comment (lajosrodek) [inline freeipa/src/test/java/com/sequenceiq/freeipa/flow/chain/UpgradeCcmFlowChainIntegrationTest.java:166]: Nit: `calledOnce` -> `calledOnceCount`. :smile:

**Review by lajosrodek — APPROVED** [review]
Only two minor bits, but approving in advance as these are absolutely not shipstoppers. Thanks for the many fixes, LGTM!
